lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,832 @@
+'use strict';
+
+const {
+  ArrayPrototypeMap,
+  PromiseAll,
+  PromisePrototypeThen,
+  PromisePrototypeFinally,
+  PromiseResolve,
+  Uint8Array,
+} = primordials;
+
+const {
+  ReadableStream,
+  isReadableStream,
+} = require('internal/webstreams/readablestream');
+
+const {
+  WritableStream,
+  isWritableStream,
+} = require('internal/webstreams/writablestream');
+
+const {
+  ByteLengthQueuingStrategy,
+  CountQueuingStrategy,
+} = require('internal/webstreams/queuingstrategies');
+
+const {
+  Writable,
+  Readable,
+  Duplex,
+} = require('stream');
+
+const {
+  Buffer,
+} = require('buffer');
+
+const {
+  errnoException,
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_STATE,
+    ERR_STREAM_PREMATURE_CLOSE,
+  },
+} = require('internal/errors');
+
+const {
+  createDeferredPromise,
+} = require('internal/util');
+
+const {
+  validateAbortSignal,
+  validateBoolean,
+  validateObject,
+} = require('internal/validators');
+
+const {
+  WriteWrap,
+  ShutdownWrap,
+  kReadBytesOrError,
+  kLastWriteWasAsync,
+  streamBaseState,
+} = internalBinding('stream_wrap');
+
+const finished = require('internal/streams/end-of-stream');
+
+const { UV_EOF } = internalBinding('uv');
+
+/**
+ * @typedef {import('../../stream').Writable} Writable
+ * @typedef {import('../../stream').Readable} Readable
+ * @typedef {import('./writablestream').WritableStream} WritableStream
+ * @typedef {import('./readablestream').ReadableStream} ReadableStream
+ *
+ * @typedef {import('../abort_controller').AbortSignal} AbortSignal
+ */
+
+/**
+ * @param {Writable} streamWritable
+ * @returns {WritableStream}
+ */
+function newWritableStreamFromStreamWritable(streamWritable) {
+  if (typeof streamWritable?._writableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamWritable',
+      'stream.Writable',
+      streamWritable);
+  }
+
```

Comment:
Should probably also handle `!streamWritable.writable`, e.g. when coming from `Duplex.toWeb`.